### PR TITLE
feat:  multiple end screens

### PR DESCRIPTION
**Pull request by Dhruwang**
## What does this PR do?

Fixes https://github.com/formbricks/internal/issues/269

Add ability to create multiple endings (endScreen or redirectToUrl)
Also contains a data migration to migrate all the existing thankyou cards endings

## How should this be tested?

Screenshot 2024-07-09 at 8 26 35 PM
Screenshot 2024-07-09 at 8 26 44 PM
Screenshot 2024-07-09 at 8 26 26 PM

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [x] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/[surveyId]/edit/components/SurveyMenuBar.tsx [hunk 1]
```diff
@@ -142,6 +142,8 @@ export const SurveyMenuBar = ({
     const localSurveyValidation = ZSurvey.safeParse(localSurvey);
     if (!localSurveyValidation.success) {
       const currentError = localSurveyValidation.error.errors[0];
+      console.log(currentError.path);
```
Comment (mattinannt) [inline apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/[surveyId]/edit/components/SurveyMenuBar.tsx:145]: please remove console log

File: apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/lib/minimalSurvey.ts [hunk 2]
```diff
@@ -23,9 +22,16 @@ export const minimalSurvey: TSurvey = {
...
   },
   questions: [],
-  thankYouCard: {
-    enabled: false,
-  },
+  endings: [
+    {
+      enabled: false,
+      type: "endScreen",
+      id: "end:1",
```
Comment (mattinannt) [inline apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/lib/minimalSurvey.ts:29]: please use cuid2 as IDs like we do it in the questions.

File: package.json [hunk 3]
```diff
@@ -43,8 +43,7 @@
   },
   "lint-staged": {
     "(apps|packages)/**/*.{js,ts,jsx,tsx}": [
-      "prettier --write",
-      "eslint --fix"
+      "prettier --write"
```
Comment (mattinannt) [inline package.json:46]: please revert change

File: packages/surveys/src/components/general/EndingCard.tsx [hunk 4]
```diff
@@ -58,10 +51,17 @@ export const ThankYouCard = ({
   );
 
   const handleSubmit = () => {
-    if (buttonLink) window.location.replace(buttonLink);
+    if (!isRedirectDisabled && endingCard.type === "endScreen" && endingCard.buttonLink) {
+      window.location.replace(endingCard.buttonLink);
```
Comment (mattinannt) [inline packages/surveys/src/components/general/EndingCard.tsx:55]: does this also needs to be `window.top?.location.replace()`?

File: packages/database/schema.prisma [hunk 5]
```diff
@@ -318,6 +318,7 @@ model Survey {
   displayPercentage  Decimal?
   languages          SurveyLanguage[]
   showLanguageSwitch Boolean?
+  endings            Json[]           @default([])
```
Comment (mattinannt) [inline packages/database/schema.prisma:321]: please move `endings` up below the `questions` so also in prisma studio this shows after the questions which makes it easier to manually make changes to records.

**Review by mattinannt — CHANGES_REQUESTED** [review]
@Dhruwang thank you for the PR 😊 It works great already 🚀, but there are still a few points we need to improve:

- Redirect Url should only be on the paid plan (at least Startup plan) for Formbricks Cloud and should be in the community edition for self-hosters. Running this locally with `IS_FORMBRICKS_CLOUD` disable I still get the following limitations even though the feature should be available:

Screenshot 2024-07-15 at 10 34 45

- When only one thank you card is created or shown, the `Delete` button (trash icon) should still be visible but disabled. In App/Website surveys also the last ending can be deleted.
Screenshot 2024-07-15 at 10 43 55

- trying to upload a title image for a thank you card throws this error:
Screenshot 2024-07-15 at 10 46 25

- The label for the redirect card shouldn't be optional. It should not be possible to save a survey without setting the label to avoid inconsistent data.
Screenshot 2024-07-15 at 10 54 52

- Recall: Can we activate recall for the url to enable dynamic redirect URLs?
Screenshot 2024-07-15 at 10 56 37

- Toggle & Navigation: The UI/UX for endings should be the same as with questions. Users should be able to duplicate endings and access the context menu (without changing the question type). For consistency we also remove the `Show` toggle as it's now possible to delete/add new endings and we don't need this anymore (this was a flaw in the concept, sorry about that)
Please make sure to use reusable components for questions and endings where it makes sense.

- Strange re-swapping behaviour: Even though changing the order works with drag and drop it looks like it doesn't since after releasing the card it goes back to it's original place (but the content changes). Please fix so it's staying in the place where it's dropped at.

https://github.com/user-attachments/assets/669e870b-a0e0-4cce-981c-cc0af333dae8

- Inconsistent behaviour in Tabs: The "End Screen" tab seems to keep the state/inputs when switching tabs and the "Redirect to URL" tab don't. Make sure to keep the state in the editor so the user can switch back and forth while editing but as soon as the changes are saved the not needed part is pruned. A survey in the database shouldn't have the type `endScreen` but a `redirectUrl` at the same time. Ideally this can be done (and validated) using zod.

https://github.com/user-attachments/assets/384bda96-0abb-4adf-a96e-e3c5fbed04c9

**Comment by jobenjada** [comment]
Looks and feels a lot nicer and cleaner :)

Left a comment on Slack regarding pot. refactoring of the very large EditEndingCard.tsx

Other than that, ready for review @mattinannt !

**Review by mattinannt — APPROVED** [review]
🚀 Ready to merge 🥳

**Pull request merged**